### ISSUE TITLE
Aerospace Window Management Integration

### ISSUE THREAD

**Issue by edheltzel** [lwouis/alt-tab-macos#3773]
**Describe the solution you'd like**
Currently, I use [Aerospace](https://github.com/nikitabobko/AeroSpace) for window management instead of the native workspaces. AltTab can show the workspace when switching apps/windows—it would be awesome if there was an option to show the Aerospace workspace as well. I'm not sure if that's possible.

**Comment by lwouis** [comment]
Sorry, it's not technically possible 🙇

**Comment by kulfix** [comment]
Meybe option to show apps only from focus workspace?

**Comment by kamaradclimber** [comment]
Would it be possible to only display windows that would be listed in a configured file?  (so that aerospace or other system would list what they consider the "relevant" windows to switch to)

**Comment by lwouis** [comment]
@kulfix this is already possible in Preferences > Controls

@kamaradclimber sorry this would see too niche usage

**Comment by kamaradclimber** [comment]
> this is already possible in Preferences > Controls

I think for aerospace the "focused" workspace is not a macos workspace (it is a notion internal to aerospace).

>  sorry this would see too niche usage
sad to hear this!

**Comment by antran22** [comment]
This is a fork somebody made that does this, however it is quite far behind now: https://github.com/XOR-op/alt-tab-macos.

**Comment by XOR-op** [comment]
> This is a fork somebody made that does this, however it is quite far behind now: https://github.com/XOR-op/alt-tab-macos.

I just made an update so it's not far behind now :)

**Comment by itpropro** [comment]
> > This is a fork somebody made that does this, however it is quite far behind now: https://github.com/XOR-op/alt-tab-macos.
> 
> I just made an update so it's not far behind now :)

@lwouis any chance this get merged back into alt-tab as aerospace option? Currently alt-tab has a lot of bugs with Aerospace, but would be a great combination considering the preview and customization features of alt-tab.

**Comment by lwouis** [comment]
@itpropro from what i can see here: https://github.com/lwouis/alt-tab-macos/compare/master...XOR-op:alt-tab-macos:7.24.0-aerospace#diff-39dcbf35d8461748414a52771fd4c3f21d5ae65aace7a204651d976b4c8fd855R5, it seems that this fork makes calls to aerospace to understand its virtual spaces.

I'm afraid we can't have AltTab be integrated so specifically with other apps. I think it would be on aerospace to make their implementation more standard. That is, accessibility apps like AltTab ideally shouldn't know aerospace is running.

Thank you 🙇

**Comment by itpropro** [comment]
I totally understand that, aerospace chose their design, because of the mess that spaces is in MacOS, so their implementation will probably not change.
It would just have been nice to have a good tab switcher available, but I get that specific integrations for tools like display managers is not in the scope of alt-tab. 
Thank you anyway for looking into this!

**Comment by patroza** [comment]
Man I came here for the very same thing.
I have the Glove80 keyboard with Sunaku's Glorious Engrammer based keymap, and it integrates so well with the Alt-Tab app (hold some key combo that opens alt-tab and then keyboard navigate it like a pro.

The only thing missing is deeper integration with AeroSpace!
Looking forward checking your fork XOR-op.

@lwouis 
In terms of how to integrate with external apps; either a generic push or pull system could work, like you could tell for a "Custom" mode that Alt-Tab should make some script call that must return a list of windows to choose from,
or the other way around an external app should be able to call Alt-Tab to say for Custom Mode, these windows should be selectable right now.

**Comment by lwouis** [comment]
[AltTab already offers a CLI interface](https://alt-tab-macos.netlify.app/#cli-usage)

**Comment by patroza** [comment]
> > This is a fork somebody made that does this, however it is quite far behind now: https://github.com/XOR-op/alt-tab-macos.
> 
> I just made an update so it's not far behind now :)

updated here to latest https://github.com/patroza/alt-tab-macos
and added support for only displaying the windows of the current focused workspace; when you enable; 
Show window from spaces: "Visible Spaces"

**Comment by epicwhale** [comment]
it would actually be quite amazing for alttab to officially support aerospace - match made in heaven! I switched workspaces to aerospace, and unfortunately the default configs were conflicting a lot with alttab, so had to uninstall alttab for now :(

**Comment by SyomaKiss** [comment]
Hey yo
This is a very needed feature. I assume I have a possible easy solution which would not require "Aerospace awareness" in Alt-Tab.
It seems like aerospace make off-screen windows to be very narrow size, like few-pixels size. 

Can we just add filter by size of windows to AltTab?
Like "only show windows wider than 10px"?

**Comment by lwouis** [comment]
We already have such a check: https://github.com/lwouis/alt-tab-macos/blob/6426d318fb73004ccb4fe809e8059ef527a3bacc/src/api-wrappers/AXUIElement.swift#L285

**Comment by SyomaKiss** [comment]
Sorry, I mislead you a lil bit. I did investigation using this command 
`./AltTab --detailed-list | python3 -m json.tool`

And what actually happens with Aerospace is that it moves the windows to position of ScreenWidth - 1px.
So what we need is to filter windows by position. 
I am already implemented this functionality and it works like we need. I did it with command line arguments which are written like this
e.g. `defaults write com.lwouis.alt-tab-macos minWindowWidth 500`

Now I am working on UI tab for "filtering", this one still does not work for me, can not update settings through it yet

**Comment by SyomaKiss** [comment]
I do not have permission to push my new branch to open PR. Should you allow it?

**Comment by SyomaKiss** [comment]
I am done, works like a charm. How should I publish my PR here?

<img width="554" height="464" alt="Image" src="https://github.com/user-attachments/assets/1b077751-30cc-401a-bcf5-ab80d3665222" />

**Comment by lwouis** [comment]
As I explained in the PR, we can't merge this feature. It would see too niche usage.

I've installed Aerospace to better understand what's going on.

So Aerospace way of managing window is to position them in the bottom-right corner. Something like 1 px to the left, so they are barely visible. I can actually see them though. It's not a great look, but I guess there was no other choice. The window are full-size. AltTab shows like sticks because that's what visible. The rest of the window is not drawn, so AltTab doesn't have the off-screen pixels to show. At least the API we use is not able to grab those.

Users can legitimately have windows be offscreen. AltTab should report that.

My conclusion is that the only way forward would be to find a way to grab the full screenshot for off-screen windows. This way we could always show them in AltTab, with full thumbnails.

That being said, it used to be like this. Apple removed the ability in Big Sur I think. The API started returned clipped screenshots. I'm afraid, nothing much we can do, if Apple is not giving us an API.

**Comment by SyomaKiss** [comment]
What you described would solve the problem of off-screen windows preview being malformed. But that is not what Aerospace users need I suppose. 
With aerospace we need option to only loop through windows within one workspace of aerospace, and this can be achieved with filtering windows by position on screen. Though you have to account for your current screen resolution, and also account for different multi-monitor setups.

Seems like from now on I have to use my own fork of the app. For anyone interested it lives here for now https://github.com/SyomaKiss/alt-tab-macos/tree/feature/windows_filtering

**Comment by XOR-op** [comment]
@SyomaKiss I would suggest you try this fix: https://github.com/XOR-op/alt-tab-macos/commit/12645a6310a02f2ea3e55c9bce91257abe92341e. Basically it prevents any thumbnail updates with any <10px dimension. Windows hidden by Aerospace will have usually have 1-2px on one dimension, so the cropped update will be ignored. As for normal AltTab users (w/o Aerospace) I don't know if that's useful enough, since the cropped size can be arbitrary.

**Comment by CannoNadav** [comment]
@SyomaKiss @lwouis I think we can work around both limitation by adding yet-another hack - if we create a virtual screen using BetterDisplay that isn't connected to any real screen and ask AeroSpace to shove all windows that aren't in the current space there then you can add an option to AltTab - "hide windows with elements inside these monitors" so the user can pick which monitors to ignore. also this will fix the annoying problem that macOS forces aerospace to have some portion of the window visible(these few visible pixels), and probably also improve AeroSpace performance because it would't need to resize windows anymore. we'll need some help from the AeroSpace team though, or at least they'll need to approve our contribution.

another option is to add an api to notify you which window Id's to ignore/blacklist. then users can add an exec-on-workspace-change callback to their aerospace.toml that notifies AltTab which windows to ignore. this option is less specific to the way that AeroSpace works.

yet-another option is to add an option to get basic window information from a "spaces server" instead of the macOS WindowServer. if anyone wants to implement a new notion of spaces then he's free to do so, he just needs to insert his server's address.

**Comment by SyomaKiss** [comment]
Hello again
Was any support for "Aerospace problem from this issue" introduced in the paid version of AltTab?

**Comment by lwouis** [comment]
Hi @SyomaKiss,
I'm not sure were Aerospace stands today. I suggest you try AltTab with Aerospace and see if it works for you. The free and Pro AltTab experiences both have the same quality regarding window detection. Pro adds more features and customization to the free experience.
Thank you

**Comment by SyomaKiss** [comment]
No, nohting has changed. The feature like Filter tab which I implemented earlier and showcased abowe with screenshot would have helped
That was fairly simple to implement, I see no reason not to add it to a paid plan, since it is "for power users" :)

**Comment by lwouis** [comment]
I think the new API we use for screenshots gets the screenshots of off-screen windows. So we should see the aerospace windows fully in the switcher now, no?

**Comment by LazyRen** [comment]
This is my one and only true wish to the AltTab.

Similar to the orginal mac-oriented workspace, Aerospace user wants `AltTab` to detect aerospace workspace, and able to filter windows from a current aerospace workspace only.

Currently this seems not possible even with the pro. With the below config, alt-tab shows all windows opened across multiple Aerospace workspaces.

<img width="741" height="530" alt="Image" src="https://github.com/user-attachments/assets/958acd62-cb24-4e29-9990-87953ba1808d" />
